### PR TITLE
[DC-631] Refactor filter code early PR

### PR DESCRIPTION
**Pull request by s-rubenstein**
Tested locally, as well as a number of unit tests on the helper methods.

Did not add a full set of render tests, because I didn't want to blow up this change with a huge amount of test data. I think we could go and add test data and render tests to most of our components, but we will want a bit more data to make that work well.
Screen Shot 2023-01-24 at 2 28 28 PM
Screen Shot 2023-01-24 at 2 28 35 PM
Screen Shot 2023-01-24 at 2 28 49 PM
Screen Shot 2023-01-24 at 2 28 53 PM
Screen Shot 2023-01-24 at 2 29 06 PM
Screen Shot 2023-01-24 at 2 29 08 PM
Screen Shot 2023-01-24 at 2 29 11 PM

### CODE REVIEW

File: src/pages/library/common.ts [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+// This function will also truncate at a word, so the beginning of the text is fully readable.
+// Example: truncateLeftWord({length: 14}, 'this is my string') -> '...my string' - the ellipses are part of the string length
+const truncateLeftWord = _.curry((options, text) => _.flow(
+  reverseText, // reverses the text so we can perform a truncate
+  _.truncate(options),
+  reverseText, // puts the text back in its original order
+  _.replace(/\.\.\.(\S+)/, '...') // Removes the first partial word
+)(text))
+
+const getContextualSuggestion = ([leftContext, match, rightContext]) => {
```
Comment (s-rubenstein) [inline src/pages/library/common.ts]: Existing code moved here.

File: src/pages/library/common.ts [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+  ])
+}
+
+const reverseText = _.flow(_.reverse, _.join(''))
+
+// truncateLeftWord
+// This will behave like Lodash's _.truncate except it will truncate from the left side of the string.
+// This function will also truncate at a word, so the beginning of the text is fully readable.
+// Example: truncateLeftWord({length: 14}, 'this is my string') -> '...my string' - the ellipses are part of the string length
+const truncateLeftWord = _.curry((options, text) => _.flow(
```
Comment (s-rubenstein) [inline src/pages/library/common.ts]: Existing code moved here

File: src/pages/library/common.ts [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+              span({ style: { opacity: 0.5, fontSize: '.75rem', lineHeight: '1.4rem' } }, [`(${numMatches})`])
+            ])
+          ])
+        ])
+      }, valuesToDisplay))
+    ])
+  ])
+}
+
+const reverseText = _.flow(_.reverse, _.join(''))
```
Comment (s-rubenstein) [inline src/pages/library/common.ts]: Existing code moved here

File: src/pages/library/common.ts [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+      filter: filter || undefined,
+      selectedSections: sections || undefined
+    }, { addQueryPrefix: true })
+
+    if (newSearch !== Nav.history.location.search) {
+      Nav.history.replace({ search: newSearch })
+    }
+  }
+
+  const onSearchChange = filter => {
```
Comment (s-rubenstein) [inline src/pages/library/common.ts]: The searching code has not really changed.

File: src/pages/library/DataBrowser.test.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  _.forEach(header => {
+    it(`generates a working matchBy function for the filter for ${header}`, () => {
+      // Arrange
+      const extractCatalogFiltersResult = extractCatalogFilters(TEST_DATASETS)
+      const section = _.find(filterSection => filterSection.header === header, extractCatalogFiltersResult)
+      if (section === undefined) {
+        fail(`Section ${header} did not appear as was expected`)
+      } else {
+        // Act
+        const matchesDatasetOne = section.matchBy(TEST_DATASET_ONE, section.values[0])
```
Comment (s-rubenstein) [inline src/pages/library/DataBrowser.test.ts]: I like the way this test is structured, but I don't like how brittle it is! If anyone has any way to have it for-each like this, while not requiring generating yet more static data, I would love to know what you come up with!

File: src/pages/library/dataBrowser-utils.test.js [hunk 6]
```diff
@@ -31,17 +34,53 @@ describe('dataBrowser-utils', () => {
...
     expect(getByText('Unspecified')).toBeTruthy()
   })
 
   it('uses given data use policy as the data use policy if unknown', () => {
-    const { getByText } = render(h(DatasetReleasePolicyDisplayInformation, { 'TerraDCAT_ap:hasDataUsePermission': 'Something else' }))
+    const { getByText } = render(makeDatasetReleasePolicyDisplayInformation('Something else'))
     expect(getByText('Something else')).toBeTruthy()
   })
+
+  it('generates consortium titles properly', () => {
```
Comment (s-rubenstein) [inline src/pages/library/dataBrowser-utils.test.js:51]: These tests could mostly get pulled into a separate PR if we wanted to make this smaller.

**Review by s-rubenstein — COMMENTED** [review]
Large diff is partially tests and partially due to the move to typescript.

File: src/pages/library/test-datasets.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import { Dataset } from 'src/libs/ajax/Catalog'
+
+
+export const TEST_DATASET_ONE: Dataset = {
+  'dct:title': 'The worlds best dataset',
+  'dct:description': 'The best there ever was',
+  'dct:creator': 'The hero we all need',
+  'dct:issued': new Date().toDateString(),
+  'dcat:accessURL': 'https://http.cat/#workspaces/',
```
Comment (s-rubenstein) [inline src/pages/library/test-datasets.ts:10]: This has /#workspaces/ in order to register as a granted dataset, rather than external

Comment (pshapiro4broad) [reply src/pages/library/test-datasets.ts:10]: Maybe add a comment here?

File: src/pages/library/dataBrowser-utils.ts [hunk 8]
```diff
@@ -1,40 +1,30 @@
...
-
-export interface Dataset extends DatasetResponse {
-  tags: Tags
-}
 
 export type DatasetAccessType =
-    'Controlled' |
     'Granted' |
     'Pending' |
+    'Controlled' |
```
Comment (pshapiro4broad) [inline src/pages/library/dataBrowser-utils.ts]: Is this because the ordering here changes the ordering in the UI? If so maybe add a comment would be good. Also I'm guessing only the ordering of one is important and the other is just to keep the two lists in sync?

File: src/pages/library/DataBrowser.test.ts (hunk 5 above)
Comment (pshapiro4broad) [reply src/pages/library/DataBrowser.test.ts]: Is is the `values[0]` vs `values[1]` that you don't like? If so, the test data package could also export an index value to avoid this hardcoding.

Or if the problem is that you don't want to hardcode asserting dataset one or two here, you could add the index to `expectedFilterSectionsAndValuesForTestDatasets` so each section could assert which dataset it expected to filter in/out.

File: src/pages/library/DataBrowser.test.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      } else {
+        fail(`Section ${sectionAndValues.header} did not appear as was expected`)
+      }
+    }, expectedFilterSectionsAndValuesForTestDatasets)
+  })
+
+  _.forEach(header => {
+    it(`generates a working matchBy function for the filter for ${header}`, () => {
+      // Arrange
+      const extractCatalogFiltersResult = extractCatalogFilters(TEST_DATASETS)
```
Comment (pshapiro4broad) [inline src/pages/library/DataBrowser.test.ts]: Can this be done outside the `forEach()`? Also it looks like this value is used in multiple tests so maybe it should only be generated once?

File: src/pages/library/DataBrowser.test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    _.forEach(expectedHeader => expect(extractCatalogFiltersResultHeaders).toContain(expectedHeader), expectedFilterHeaders)
+  })
+
+  it('generates the correct values for each filter section', () => {
+    // Act
+    const extractCatalogFiltersResult = extractCatalogFilters(TEST_DATASETS)
+    // Assert
+    _.forEach(sectionAndValues => {
+      const filter = _.find(filter => filter.header === sectionAndValues.header, extractCatalogFiltersResult)
+      if (filter !== undefined) {
```
Comment (pshapiro4broad) [inline src/pages/library/DataBrowser.test.ts:100]: Is this the usual convention for the UI tests to check for `undefined` explicitly? To me this seems equivalent to a java test that checks for `null` before asserting on the contents of an object. IMO I wouldn't have a separate assert in java as the dereference will trigger an NPE so there's no need for an explicit check; and because I prefer having less test code to maintain vs test code that has explicit checks to provide nicer assert messages.

Or maybe in this case the explicit check is needed because without it the test won't fail?

File: src/pages/library/SearchAndFilterComponent.test.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  })
+
+  it('matches on intersection for a list of sections', () => {
+    // Act
+    const listItemsMatching = getMatchingDataForSectionList(FILTERS, data)
+    // Assert
+    expect(listItemsMatching).not.toContain(data[0])
+    expect(listItemsMatching).not.toContain(data[1])
+    expect(listItemsMatching).not.toContain(data[2])
+    expect(listItemsMatching).not.toContain(data[3])
```
Comment (pshapiro4broad) [inline src/pages/library/SearchAndFilterComponent.test.ts]: Maybe add `xylophone` to your data to add a positive test here?

**Review by pshapiro4broad — APPROVED** [review]
Looks good, only a few comments

File: src/pages/library/DataBrowser.test.ts (hunk 10 above)
Comment (s-rubenstein) [reply src/pages/library/DataBrowser.test.ts:100]: It is needed because typescript complains about calling methods on potentially undefined types (without it the type is `FilterSection | undefined`. So it provides a compiletime error. This check ensures the typing is correct. I also tried to do it with a cast, but the cast didn't work.

File: src/pages/library/dataBrowser-utils.ts (hunk 8 above)
Comment (s-rubenstein) [reply src/pages/library/dataBrowser-utils.ts]: The ordering does change the ordering in the UI, but I am actually going to revert this change and change the test that depended on the order to not.

File: src/pages/library/dataBrowser-utils.ts [hunk 12]
```diff
@@ -161,32 +151,12 @@ export const DatasetAccess = ({ dataset }: DatasetAccessProps) => {
...
-      dataset['TerraDCAT_ap:hasDataUsePermission']
-    ]) as string
-  }
-}
-
 
-export const prepareDatasetsForDisplay = (datasets: DatasetResponse[], dataCollectionsToInclude: string[]): Dataset[] => {
-  const filteredDatasets = _.filter(dataCollectionsToInclude ?
+export const prepareDatasetsForDisplay = (datasets: Dataset[], dataCollectionsToInclude: string[]): Dataset[] => {
+  return _.filter(dataCollectionsToInclude ?
```
Comment (fboulnois) [inline src/pages/library/dataBrowser-utils.ts]: Can we avoid `return`ing here directly by doing `... => _.filter(dataCollectionsToInclude ? ...`

File: src/pages/library/dataBrowser-utils.ts [hunk 13]
```diff
@@ -57,37 +47,37 @@ export const getDatasetReleasePoliciesDisplayInformation = (dataUsePermission: s
...
+export const isWorkspace = (dataset: Dataset): boolean => {
   return _.toLower(dataset['dcat:accessURL']).includes(workspaceUrlFragment)
 }
 
 export const datarepoSnapshotUrlFragment = '/snapshots/details/'
 
-export const isDatarepoSnapshot = (dataset: DatasetResponse) => {
+export const isDatarepoSnapshot = (dataset: Dataset): boolean => {
   return _.toLower(dataset['dcat:accessURL']).includes(datarepoSnapshotUrlFragment)
 }
```
Comment (fboulnois) [inline src/pages/library/dataBrowser-utils.ts]:
```suggestion
export const isDatarepoSnapshot = (dataset: Dataset): boolean => _.toLower(dataset['dcat:accessURL']).includes(datarepoSnapshotUrlFragment)
```

Ditto here?

File: src/pages/library/dataBrowser-utils.ts [hunk 14]
```diff
@@ -57,37 +47,37 @@ export const getDatasetReleasePoliciesDisplayInformation = (dataUsePermission: s
...
   [isDatarepoSnapshot(dataset), () => false],
-  () => true) as boolean
+  () => true)
 
 export const workspaceUrlFragment = '/#workspaces/'
 
-export const isWorkspace = (dataset: DatasetResponse) => {
+export const isWorkspace = (dataset: Dataset): boolean => {
   return _.toLower(dataset['dcat:accessURL']).includes(workspaceUrlFragment)
 }
```
Comment (fboulnois) [inline src/pages/library/dataBrowser-utils.ts]:
```suggestion
export const isWorkspace = (dataset: Dataset): boolean => _.toLower(dataset['dcat:accessURL']).includes(workspaceUrlFragment)
```

Ditto here?

File: src/pages/library/dataBrowser-utils.ts [hunk 15]
```diff
@@ -44,7 +34,7 @@ export const uiMessaging = {
...
 
 // This list is generated from the schema enum
-export const getDatasetReleasePoliciesDisplayInformation = (dataUsePermission: string): { label: string; description?: string } => {
+export const getDatasetReleasePoliciesDisplayInformation = (dataUsePermission?: string): { label: string; description?: string } => {
   return Utils.switchCase(
```
Comment (fboulnois) [inline src/pages/library/dataBrowser-utils.ts]:
```suggestion
export const getDatasetReleasePoliciesDisplayInformation = (dataUsePermission?: string): { label: string; description?: string } => Utils.switchCase(
```

Ditto here?

File: src/pages/library/dataBrowser-utils.test.js (hunk 6 above)
Comment (fboulnois) [reply src/pages/library/dataBrowser-utils.test.js:51]: I think it's worth keeping them in here since we have coverage on each PR now.

File: src/pages/library/SearchAndFilterComponent.test.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+describe('library/common', () => {
+  it('matches correctly for a section entry', () => {
+    // Act
+    const listItemsMatching = listItemsMatchForSectionEntry('a', FILTERS[0].matchBy, data)
+    // Assert
+    expect(listItemsMatching).toContain(data[0])
+    expect(listItemsMatching).toContain(data[1])
+    expect(listItemsMatching).not.toContain(data[2])
+    expect(listItemsMatching).not.toContain(data[3])
+    expect(listItemsMatching).not.toContain(data[4])
```
Comment (fboulnois) [inline src/pages/library/SearchAndFilterComponent.test.ts]: Would it be worth making `data` an enum so that it's clear which of these items is being tested across these tests? For example `expect(listItemsMatching).not.toContain(data.STOCK)`

**Review by fboulnois — COMMENTED** [review]
A few suggestions:

**Comment by fboulnois** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=DataBiosphere_terra-ui&pullRequest=3721)
> 
> [Bug](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=DataBiosphere_terra-ui&pullRequest=3721&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=DataBiosphere_terra-ui&pullRequest=3721&metric=new_coverage&view=list) [40.2% Coverage](https://sonarcloud.io/component_measures?id=DataBiosphere_terra-ui&pullRequest=3721&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=DataBiosphere_terra-ui&pullRequest=3721&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=DataBiosphere_terra-ui&pullRequest=3721&metric=new_duplicated_lines_density&view=list)

It's also worth taking a look a the code smell and seeing if there are obvious coverage issues that could be fixed (otherwise I think it's worth adding a ticket for additional unit tests)

**Comment by s-rubenstein** [comment]
I looked at the code smell, it is largely ternary operations and default values using `|| ''` that we have as established patterns in the codebase. I added discussion about it to the UIWG on Tuesday.

**Review by fboulnois — APPROVED** [review]
lgtm
